File: backend/app/services/appointment_service.py
"""
Appointment service — business logic for appointment management.
"""

import math
from datetime import datetime, timezone
from typing import Optional

from fastapi import HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.appointment import Appointment, AppointmentStatus
from app.models.user import User
from app.repositories.appointment_repository import AppointmentRepository
from app.schemas.appointment import (
    AppointmentCreate,
    AppointmentUpdate,
    AppointmentResponse,
    AppointmentListResponse,
)


class AppointmentService:
    def __init__(self, db: AsyncSession):
        self.repo = AppointmentRepository(db)

    async def get_appointments(
        self,
        page: int = 1,
        size: int = 20,
        status_filter: Optional[AppointmentStatus] = None,
        patient_id: Optional[int] = None,
    ) -> AppointmentListResponse:

        skip = (page - 1) * size

        items = await self.repo.get_all_with_relations(
            skip=skip,
            limit=size,
            status=status_filter,
            patient_id=patient_id,
        )

        total = await self.repo.count_filtered(
            status=status_filter,
            patient_id=patient_id,
        )

        pages = math.ceil(total / size) if total > 0 else 1

        return AppointmentListResponse(
            items=[
                AppointmentResponse.model_validate(a)
                for a in items
            ],
            total=total,
            page=page,
            size=size,
            pages=pages,
        )

    async def get_appointment(
        self,
        appointment_id: int,
    ) -> Appointment:

        appt = await self.repo.get_by_id_with_relations(
            appointment_id
        )

        if not appt:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Jadwal dengan ID {appointment_id} tidak ditemukan",
            )

        return appt

    async def create_appointment(
        self,
        data: AppointmentCreate,
        created_by: User,
    ) -> Appointment:

        # Generate appointment number
        number = await self.repo.generate_appointment_number()

        # Create appointment object
        appt = Appointment(
            **data.model_dump(),
            appointment_number=number,
            created_by_id=created_by.id,
            status=AppointmentStatus.PENDING,
        )

        # Save to database
        result = await self.repo.create(appt)

        # Reload with relationships
        appointment = await self.repo.get_by_id_with_relations(
            result.id
        )

        if not appointment:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Gagal memuat appointment setelah dibuat",
            )

        return appointment

    async def update_appointment(
        self,
        appointment_id: int,
        data: AppointmentUpdate,
    ) -> Appointment:

        appt = await self.get_appointment(
            appointment_id
        )

        update_data = data.model_dump(
            exclude_none=True
        )

        new_status = update_data.get("status")

        # Validate status transition
        if new_status:
            self._validate_status_transition(
                appt.status,
                new_status,
            )

        # Auto set checked_in_at
        if (
            new_status == AppointmentStatus.CHECKED_IN
            and not update_data.get("checked_in_at")
        ):
            update_data["checked_in_at"] = datetime.now(
                timezone.utc
            )

        updated = await self.repo.update(
            appt,
            update_data,
        )

        # Reload relationships
        appointment = await self.repo.get_by_id_with_relations(
            updated.id
        )

        return appointment

    async def cancel_appointment(
        self,
        appointment_id: int,
        reason: Optional[str] = None,
    ) -> Appointment:

        appt = await self.get_appointment(
            appointment_id
        )

        if appt.status in [
            AppointmentStatus.COMPLETED,
            AppointmentStatus.CANCELLED,
        ]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=(
                    f"Jadwal dengan status "
                    f"'{appt.status}' tidak dapat dibatalkan"
                ),
            )

        updated = await self.repo.update(
            appt,
            {
                "status": AppointmentStatus.CANCELLED,
                "cancellation_reason": reason,
            },
        )

        # Reload relationships
        appointment = await self.repo.get_by_id_with_relations(
            updated.id
        )

        return appointment

    def _validate_status_transition(
        self,
        current: AppointmentStatus,
        new: AppointmentStatus,
    ) -> None:

        allowed: dict[
            AppointmentStatus,
            list[AppointmentStatus]
        ] = {
            AppointmentStatus.PENDING: [
                AppointmentStatus.CONFIRMED,
                AppointmentStatus.CANCELLED,
            ],
            AppointmentStatus.CONFIRMED: [
                AppointmentStatus.CHECKED_IN,
                AppointmentStatus.CANCELLED,
                AppointmentStatus.NO_SHOW,
            ],
            AppointmentStatus.CHECKED_IN: [
                AppointmentStatus.COMPLETED,
                AppointmentStatus.CANCELLED,
            ],
            AppointmentStatus.COMPLETED: [],
            AppointmentStatus.CANCELLED: [],
            AppointmentStatus.NO_SHOW: [],
        }

        if new not in allowed.get(current, []):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=(
                    f"Tidak dapat mengubah status "
                    f"dari '{current}' ke '{new}'"
                ),
            )

File: frontend/src/services/appointmentService.ts
import api from "./api";

export type AppointmentStatus = "pending" | "confirmed" | "checked_in" | "completed" | "cancelled" | "no_show";
export type AppointmentPriority = "routine" | "urgent" | "emergency";

export interface Appointment {
  id: number;
  appointment_number: string;
  patient_id: number;
  referring_doctor_id: number | null;
  created_by_id: number | null;
  requested_modality: string;
  body_part: string;
  clinical_indication: string | null;
  preparation_instructions: string | null;
  scheduled_datetime: string;
  estimated_duration_minutes: number;
  status: AppointmentStatus;
  priority: AppointmentPriority;
  notes: string | null;
  cancellation_reason: string | null;
  checked_in_at: string | null;
  patient: { id: number; full_name: string; medical_record_number: string } | null;
  referring_doctor: { id: number; full_name: string; role: string } | null;
  created_at: string;
  updated_at: string;
}

export interface AppointmentListResponse {
  items: Appointment[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface CreateAppointmentData {
  patient_id: number;
  requested_modality: string;
  body_part: string;
  scheduled_datetime: string;
  clinical_indication?: string;
  preparation_instructions?: string;
  estimated_duration_minutes?: number;
  priority?: AppointmentPriority;
  notes?: string;
  referring_doctor_id?: number;
}

export const appointmentService = {
  async list(params?: { page?: number; size?: number; status?: string; patient_id?: number }): Promise<AppointmentListResponse> {
    const { data } = await api.get("/api/v1/appointments", { params });
    return data;
  },
  async get(id: number): Promise<Appointment> {
    const { data } = await api.get(`/api/v1/appointments/${id}`);
    return data;
  },
  async create(payload: CreateAppointmentData): Promise<Appointment> {
    const { data } = await api.post("/api/v1/appointments", payload);
    return data;
  },
  async update(id: number, payload: Partial<Appointment>): Promise<Appointment> {
    const { data } = await api.patch(`/api/v1/appointments/${id}`, payload);
    return data;
  },
  async cancel(id: number, reason?: string): Promise<Appointment> {
    const { data } = await api.post(`/api/v1/appointments/${id}/cancel`, null, {
      params: reason ? { reason } : undefined,
    });
    return data;
  },
};
